src/datastore/datastore.rs

```diff
@@ -261,7 +261,12 @@ impl DatastoreInstance {
             Err(err) => return Err(DatastoreError::InternalError(format!("Failed to prepare get_stored_buckets SQL statement: {}", err.to_string())))
         };
         let buckets = match stmt.query_map(&[] as &[&dyn ToSql], |row| {
-            let data_str : String = row.get(6)?;
+            // If data column is not set (possible on old installations), use an empty map as default
```

Comment:
This is not how it should be solved as it creates ugly exceptions for legacy code.

Higher up in the _create_tables function you can see this:

        conn.execute("ALTER TABLE buckets ADD COLUMN data TEXT default '{}';", &[] as &[&dyn ToSql])

To me it seems like this should work, but apparently it does not? That's what should be fixed.

Reply:
Huh, that's weird. Works fine when I try the equivalent commands on in an sqlite3 REPL.

```
sqlite> create table buckets (id INTEGER PRIMARY KEY, name TEXT);
sqlite> .table
buckets
sqlite> insert into buckets (name) values ("test");
sqlite> select * from buckets;
1|test
sqlite> ALTER TABLE buckets ADD COLUMN data TEXT default '{}';
sqlite> select * from buckets;
1|test|{}
sqlite> PRAGMA user_version;
0
sqlite> PRAGMA user_version = 2;
sqlite> PRAGMA user_version;
2
```

Reply:
Well, I don't know how to reproduce so I can't really dig into it. It's the classical case of "works fine on my machine".

Reply:
Well, after a lot of testing I've had no trouble reproducing. It happens every time on Android, **even on a fresh install**, but never when run on my laptop.

I suggest we merge this and call it a day, I really have no idea where else to go from here, unless you have an idea.

Reply:
The thing is, the code doesn't even allow "data" to contain null since the beginning of June when I merged my large PR, only a Map<String, Value> so it doesn't make any sense.

https://github.com/ActivityWatch/aw-server-rust/blob/master/src/models/bucket.rs#L21

I really want to avoid having to maintain legacy code like this.

Reply:
Alright, apparently Android doesn't clear app data on an uninstall. After clearing data and running again it did not occur.

Reply:
>The thing is, the code doesn't even allow "data" to contain null since the beginning of June when I merged my large PR, only a Map<String, Value> so it doesn't make any sense.

It's not `Bucket.data` that contains `null`, it's the sqlite field that contains the string `"null"`.

Reply:
People seem to suggest one should always use `NOT NULL` with `TEXT DEFAULT`: https://stackoverflow.com/a/6599261/965332

Why it doesn't respect the `DEFAULT` specifically on Android I don't know (could be due to how it's called from the code, leading to `NULL`s appearing) but it would obviously solve the problem in our case.

Now the question is: How do we migrate from a `TEXT DEFAULT` data column to a `TEXT NOT NULL DEFAULT` data column?

Reply:
3. During migration:
 - do a select on data for all rows and save them in memory
 - delete the data column
 - create it again but now with NOT NULL
 - Restore the saved data fields, If the old data was NULL don't set it as that will make it '{}' which should be correct

Reply:
You cannot delete a column.

Reply:
I added a migration that sort of solved it: https://github.com/ActivityWatch/aw-server-rust/pull/52/commits/2b55a97549429c4b570b28bd8d05396a22d8046e

I've already spent way too much time (5h+) on what is practically nitpicking, so if this works I will merge it.